secboot/keymgr/keymgr_luks2.go

```diff
@@ -0,0 +1,212 @@
+	// key slot used by the recovery key
+	recoveryKeySlot = 1
+	// temporary key slot used when changing the encryption key
+	tempKeySlot = recoveryKeySlot + 1
+)
+
+var (
+	sbGetDiskUnlockKeyFromKernel = sb.GetDiskUnlockKeyFromKernel
+	keyringAddKeyToUserKeyring   = keyring.AddKeyToUserKeyring
+)
+
+func getEncryptionKeyFromUserKeyring(dev string) ([]byte, error) {
+	const remove = false
+	const defaultPrefix = "ubuntu-fde"
+	// note this is the unlock key, which can be either the main key which
+	// was unsealed, or the recovery key, in which case some operations may
+	// not make sense
+	currKey, err := sbGetDiskUnlockKeyFromKernel(defaultPrefix, dev, remove)
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain current unlock key for %v: %v", dev, err)
+	}
+	return currKey, err
+}
+
+func isKeyslotNotActive(err error) bool {
+	match, _ := regexp.MatchString(`.*: Keyslot [0-9]+ is not active`, err.Error())
+	return match
+}
+
+func recoveryKDF() (*luks2.KDFOptions, error) {
+	usableMem, err := osutil.TotalUsableMemory()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get usable memory for KDF parameters when adding the recovery key: %v", err)
+	}
+	// The KDF memory is heuristically calculated by taking the
+	// usable memory and subtracting hardcoded 384MB that is
+	// needed to keep the system working. Half of that is the mem
+	// we want to use for the KDF. Doing it this way avoids the expensive
+	// benchmark from cryptsetup. The recovery key is already 128bit
+	// strong so we don't need to be super precise here.
+	kdfMem := (int(usableMem) - 384*1024*1024) / 2
+	// max 1 GB
+	if kdfMem > 1024*1024*1024 {
+		kdfMem = (1024 * 1024 * 1024)
+	}
+	// min 32 KB
+	if kdfMem < 32*1024 {
+		kdfMem = 32 * 1024
+	}
+	return &luks2.KDFOptions{
+		MemoryKiB:       kdfMem / 1024,
+		ForceIterations: 4,
+	}, nil
+}
+
+// AddRecoveryKeyToLUKSDevice adds a recovery key to a LUKS2 device. It the
+// devuce unlock key from the user keyring to authorize the change. The
+// recoveyry key is added to keyslot 1.
+func AddRecoveryKeyToLUKSDevice(recoveryKey keys.RecoveryKey, dev string) error {
+	currKey, err := getEncryptionKeyFromUserKeyring(dev)
+	if err != nil {
+		return err
+	}
+
+	return AddRecoveryKeyToLUKSDeviceUsingKey(recoveryKey, currKey, dev)
+}
+
+// AddRecoveryKeyToLUKSDeviceUsingKey adds a recovery key rkey to the existing
+// LUKS encrypted volume on the block device given by node. The existing key to
+// the encrypted volume is provided in the key argument and used to authorize
+// the operation.
+//
+// A heuristic memory cost is used.
+func AddRecoveryKeyToLUKSDeviceUsingKey(recoveryKey keys.RecoveryKey, currKey keys.EncryptionKey, dev string) error {
+	opts, err := recoveryKDF()
+	if err != nil {
+		return err
+	}
+
+	options := luks2.AddKeyOptions{
+		KDFOptions: *opts,
+		Slot:       recoveryKeySlot,
+	}
+	if err := luks2.AddKey(dev, currKey, recoveryKey[:], &options); err != nil {
+		return fmt.Errorf("cannot add key: %w", err)
```

Comment:
I've nothing against `%w`, but unless there's a precise reason to use them here (in other parts of the same file you use `%v`) I'd stick with `%v` until we take a decision to start using `%w` in newly written code. If, on the other hand, there's a reason to prefer `%w` here, I'd add a comment to make sure no one later changes this back to a `%v`.